jesseduffield/lazygit
# Rebasing/Cherry-Picking/Mouse-Support

This is the continuation of @glvr182's work that I picked up and then didn't get around to for ages (sorry :3)


![image](https://user-images.githubusercontent.com/8456633/53295911-866c4c80-3859-11e9-8e4d-387818c952bc.png)
Interactive rebase onto commits or other branches, with the ability to easy edit/squash/reword/drop/fixup/rearrange commits without needing to edit the TODO file manually.

![image](https://user-images.githubusercontent.com/8456633/53295913-9421d200-3859-11e9-9180-5ac710ec976a.png)
Cherry pick commits by coping them on one branch and pasting them onto another

This PR includes a bit of rework that needed to be done to support this feature.
In lieu of windows (i.e. objects containing a set of views that can be shown at different times depending on which 'tab' you're on) I've introduced a concept of 'contexts' where you can have a view that has different keybindings applying to it depending on the context. In this case the main view (the one on the right) can either be in the 'normal', 'staging', or 'merging' contexts.

I've also changes the logic for loaders because previously whenever an error message from git contained an ellipsis (...) it would chuck a loader symbol next to it and hide the rest of the message.

I've also added a hook for focus changing, which kind of awkwardly co-exists with the switchFocus method from view_helpers.go but I'm anticipating a big refactor that will clean that up.

The way this feature works is that when you hit 'r' on a branch in the branches panel, you get asked if you want to rebase, and if so, it begins the rebase against that branch's head. If there are conflicts, you're given a commiserative message with the option to hit 'esc' to abort or 'enter' to begin resolving merge conflicts. Then either from your editor or from the merge panel (by hitting space/enter on a file and picking hunks with space) you can resolve the conflicts. 

If you've just resolved the last conflict from within lazygit, you're given a prompt to continue the rebase. If you continue and there are more conflicts then the process repeats.

At any time, in the files panel you can hit 'm' to view the three merge/rebase options (abort/skip/continue).

By default, merge commits will be made automatically, but if you are on a unix machine, you can set the `git.merging.manualCommit` setting to true and your editor will open when you need to make the merge commit.

To test:
in `test/repos/repo` (mkdir repo if it doesn't exist there), you can use 
```
cd ..; ./merge_conflict.sh; cd repo
```
 to give you a fresh repo that's ready for the develop branch to be merged into the master branch (or rebased). There's a few pathways worth testing: for example:
1) only picking the top hunks in each of the conflict files (this results in a git merge --skip rather than a git merge --continue)
2) aborting when the prompt gives you the option
3) trying to merge while in the middle of a rebase
4) seeing what happens after you get merge conflicts after applying a stash entry (You should not see any attempt at continuing a merge once you resolve the final conflict)
